src/Pim/Component/Catalog/Factory/ProductValueFactory.php

```
@@ -18,50 +19,41 @@ class ProductValueFactory
+     * @param ProductValueFactoryRegistry $registry
      */
     public function __construct(
         AttributeValidatorHelper $attributeValidatorHelper,
-        $productValueClass
+        ProductValueFactoryRegistry $registry
     ) {
-        if (!class_exists($productValueClass)) {
-            throw new \InvalidArgumentException(
-                sprintf('The product value class "%s" does not exist.', $productValueClass)
-            );
-        }
-
-        $this->productValueClass = $productValueClass;
+        $this->registry = $registry;
         $this->attributeValidatorHelper = $attributeValidatorHelper;
     }
 
     /**
-     * This method effectively creates an empty product value while checking the provided localeCode and ChannelCode
-     * exists.
-     * The Data for this product value should be set in a second time using ProductValue::setData method.
+     * This method effectively creates a product value and set its data, while
+     * checking the provided localeCode and ChannelCode exists.
      *
      * @param AttributeInterface $attribute
      * @param string             $channelCode
      * @param string             $localeCode
+     * @param mixed              $data
      *
      * @throws \LogicException
      *
      * @return ProductValueInterface
      */
-    public function create(AttributeInterface $attribute, $channelCode, $localeCode)
+    public function create(AttributeInterface $attribute, $channelCode, $localeCode, $data)
     {
         $this->attributeValidatorHelper->validateScope($attribute, $channelCode);
         $this->attributeValidatorHelper->validateLocale($attribute, $localeCode);
 
-        /** @var ProductValueInterface $value */
-        $value = new $this->productValueClass();
-        $value->setAttribute($attribute);
-        $value->setScope($channelCode);
-        $value->setLocale($localeCode);
+        $factory = $this->registry->get($attribute->getAttributeType());
```

what if get returns nothing?
we should throw an exception here or directly in the registry

---

You already did it in the registry ;)